### ISSUE TITLE
Questions for Rate Limiting

### ISSUE THREAD

**Issue by markjiang7m2** [ThreeMammals/Ocelot#910]
## Expected Behavior
I setup a demo for Ocelot Rate Limiting function. Period: 1m, PeriodTimespan: 10, Limit: 5. As per the [Documentation for Rate Limiting](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html), The count of request should be limitted as 5 in 1 min, and then after 10 seconds, it will works again.  

## Actual Behavior
I setup a console project for sending the request. It will sleep 1s after got a response. Below is the output: 
```
2019/6/3 13:33:31
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:32
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:33
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:34
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:35
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:36
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
2019/6/3 13:33:37
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
2019/6/3 13:33:38
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
2019/6/3 13:33:39
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
2019/6/3 13:33:40
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
2019/6/3 13:33:41
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:43
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:44
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:45
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:46
OK, This is from localhost:8001, path: /api/ocelot/5
2019/6/3 13:33:47
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
2019/6/3 13:33:48
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
2019/6/3 13:33:49
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
2019/6/3 13:33:50
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
2019/6/3 13:33:51
TooManyRequests, API calls quota exceeded! maximum admitted 5 per 1m.
```
It is 5 successes, 5 rejects as period.  
Then I change the RateLimitOptions as Period: 1m, PeriodTimespan: 10, Limit: 6, and I got 6 successes, 4 rejects as period.

## My Questions  
1. Does the time of last reject = The time of the 1st success + PeriodTimespan?  
2. After the PeriodTimespan, request will works again, but in a Period(1min here), the count of successful requests is more than the limit(5 times here). What is the logic between Period and PeriodTimespan? Is there anything wrong with my setting?

## Steps to Reproduce the Problem
Api service:  
```c#
[HttpGet("{id}")]
public async Task<IActionResult> Get(int id)
{
	var result = await Task.Run(() =>
	{
		return $"This is from {HttpContext.Request.Host.Value}, path: {HttpContext.Request.Path}";
	});
	return Ok(result);
}
```

Ocelot.json:  
```js
{
  "DownstreamPathTemplate": "/api/ocelot/{postId}",
  "DownstreamScheme": "http",
  "DownstreamHostAndPorts": [
	{
	  "Host": "localhost",
	  "Port": 8001
	}
  ],
  "UpstreamPathTemplate": "/ocelot/ratelimit/{postId}",
  "UpstreamHttpMethod": [ "Get" ],
  "Priority": 2,
  "RateLimitOptions": {
	"ClientWhiteList": [
	  "markadmin"
	],
	"EnableRateLimiting": true,
	"Period": "1m",
	"PeriodTimespan": 10,
	"Limit": 5
  }
}
```

Console project:  
```c#
public static async Task Main(string[] args)
{
	using (var client = new HttpClient())
	{
		for (var i = 0; i < 100; i++)
		{
			Console.WriteLine(DateTime.Now);

			var result = await client.GetAsync("http://localhost:4727/ocelot/ratelimit/5");
			Console.WriteLine($"{result.StatusCode}, {result.Content.ReadAsStringAsync().Result}");
			System.Threading.Thread.Sleep(1000);
		}
		Console.Read();
	}
}
```

## Specifications

  - Version:13.5.1
  - Platform:.net core 2.2.0
  - Subsystem:

**Comment by markjiang7m2** [comment]
One more question. I am trying to use the LoadBalancer in this demo, so I change the reroute as below.  
```js
{
  "DownstreamPathTemplate": "/api/ocelot/{postId}",
  "DownstreamScheme": "http",
  "DownstreamHostAndPorts": [
	{
	  "Host": "localhost",
	  "Port": 8001
	},
        {
	  "Host": "localhost",
	  "Port": 8002
	}
  ],
  "UpstreamPathTemplate": "/ocelot/ratelimit/{postId}",
  "UpstreamHttpMethod": [ "Get" ],
  "Priority": 2,
  "RateLimitOptions": {
	"ClientWhiteList": [
	  "markadmin"
	],
	"EnableRateLimiting": true,
	"Period": "1m",
	"PeriodTimespan": 10,
	"Limit": 5
  },
    "LoadBalancerOptions": {
    "Type": "RoundRobin"
    }
}
```

As the document say that  
`RoundRobin - loops through available services and sends requests. `  

I do the test as below.  
Firstly, both downstream services are running correctly, and the results are correct. Then I try stopping the `8001` service, Ocelot will still transfer request to `8001` service, so I get a half bad request results. Is it correct?

**Comment by ahmed0304** [comment]
@markjiang7m2, can you please tell me from where you are getting `markadmin` in `RateLimitOptions:ClientWhiteList` ?

**Comment by raman-m** [comment]
>@markjiang7m2 [on Jun 3, 2019](https://github.com/ThreeMammals/Ocelot/issues/910#issue-451322615)

Hi Mark!
Thanks for bug reporting!

Current [Rate Limiting issues](https://github.com/ThreeMammals/Ocelot/issues?q=is%3Aopen+is%3Aissue+label%3A%22Rate+Limiting%22)
We have #1590 which seems is very similar to your user scenario. I would say #1590 is the duplicate of this issue. Please confirm that fact!

So, I'll keep this issue open, to let it be fixed together with #1590...

**Comment by raman-m** [comment]
> @markjiang7m2 commented [on Jun 5, 2019](https://github.com/ThreeMammals/Ocelot/issues/910#issuecomment-498921457)

Regarding Load Balancer and **RoundRobin**...
That is already known bug #1513 which we accepted. Seems your case is related to #1041 also.

**Comment by raman-m** [comment]
Rate limiting bug was fixed in #1590 
Bug fix is available in versions [23.3.0](https://github.com/ThreeMammals/Ocelot/releases/tag/23.3.0)+
